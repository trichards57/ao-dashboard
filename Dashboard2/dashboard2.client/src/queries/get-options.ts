import { queryOptions } from "@tanstack/react-query";
import { notFound } from "@tanstack/react-router";

export default function getOptions<T>(uri: string, queryKey: string[]) {
  return queryOptions({
    queryKey,
    queryFn: async () => {
      const response = await fetch(uri, {
        method: "GET",
        headers: {
          "Content-Type": "application/json",
        },
      });

      if (!response.ok) {
        if (response.status === 404) {
          throw notFound();
        }
        throw new Error("Network error");
      }

      return response.json() as Promise<T>;
    },
    staleTime: 10 * 60 * 1000,
    throwOnError: true,
  });
}
