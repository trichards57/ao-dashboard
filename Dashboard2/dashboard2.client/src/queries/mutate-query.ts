import { useMutation, useQueryClient } from "@tanstack/react-query";
import { notFound } from "@tanstack/react-router";

export default function useUpdate(uri: string, queryKey: string[]) {
  const queryClient = useQueryClient();

  return function useUpdateInner<TBody>(id: string) {
    return useMutation({
      mutationFn: async (body: TBody) => {
        const response = await fetch(`${uri}/${id}`, {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify(body),
        });

        if (!response.ok) {
          if (response.status === 404) {
            // eslint-disable-next-line @typescript-eslint/no-throw-literal
            throw notFound();
          }
          throw new Error("Failed to update.");
        }
      },
      onSuccess: () => {
        queryClient.invalidateQueries({ queryKey });
      },
    });
  };
}
