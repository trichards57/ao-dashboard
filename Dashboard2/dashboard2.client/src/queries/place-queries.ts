import {
  QueryClient,
  queryOptions,
  useSuspenseQuery,
} from "@tanstack/react-query";
import { notFound } from "@tanstack/react-router";

export type Region =
  | "All"
  | "EastOfEngland"
  | "EastMidlands"
  | "London"
  | "NorthEast"
  | "NorthWest"
  | "SouthEast"
  | "SouthWest"
  | "WestMidlands";

export function districtOptions(region: Region) {
  return queryOptions({
    queryKey: ["districts", region],
    queryFn: async () => {
      if (region == "All") {
        return [];
      }

      const response = await fetch(`/api/places/${region}`, {
        method: "GET",
        headers: {
          "Content-Type": "application/json",
        },
      });

      if (!response.ok) {
        if (response.status === 404) {
          throw notFound();
        }
        throw new Error("Failed to fetch districts.");
      }

      return response.json() as Promise<string[]>;
    },
    staleTime: 10 * 60 * 1000,
  });
}

export function hubOptions(region: Region, district: string) {
  return queryOptions({
    queryKey: ["hubs", region, district],
    queryFn: async () => {
      if (region == "All" || district.toUpperCase() == "ALL") {
        return [];
      }

      const response = await fetch(`/api/places/${region}/${district}`, {
        method: "GET",
        headers: {
          "Content-Type": "application/json",
        },
      });

      if (!response.ok) {
        if (response.status === 404) {
          throw notFound();
        }
        throw new Error("Failed to fetch hubs.");
      }

      return response.json() as Promise<string[]>;
    },
    staleTime: 10 * 60 * 1000,
  });
}

export function useDistricts(region: Region) {
  return useSuspenseQuery(districtOptions(region));
}

export function useHubs(region: Region, district: string) {
  return useSuspenseQuery(hubOptions(region, district));
}

export function preloadDistricts(queryClient: QueryClient, region: Region) {
  return queryClient.ensureQueryData(districtOptions(region));
}

export function preloadHubs(
  queryClient: QueryClient,
  region: Region,
  district: string,
) {
  return queryClient.ensureQueryData(hubOptions(region, district));
}
