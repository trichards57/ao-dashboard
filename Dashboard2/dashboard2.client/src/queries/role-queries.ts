import { QueryClient, useSuspenseQuery } from "@tanstack/react-query";

import getOptions from "./get-options";
import useUpdate from "./mutate-query";

export type Permission = "Deny" | "Read" | "Write";

export interface RolePermissions {
  id: string;
  name: string;
  permissions: Permission;
  vehicleConfiguration: Permission;
  vorData: Permission;
}

export interface RolePermissionsUpdate {
  permissions: Permission;
  vehicleConfiguration: Permission;
  vorData: Permission;
}

export function useUpdateRole(id: string) {
  return useUpdate("/api/roles", ["role"])<RolePermissionsUpdate>(id);
}

export function roleOptions(id: string) {
  return getOptions<RolePermissions>(`/api/roles/${id}`, ["role", id]);
}

export const allRoleOptions = getOptions<RolePermissions[]>("/api/roles", [
  "role",
]);

export function useRole(id: string) {
  return useSuspenseQuery(roleOptions(id));
}

export function useRoles() {
  return useSuspenseQuery(allRoleOptions);
}

export function preloadRole(queryClient: QueryClient, id: string) {
  return queryClient.ensureQueryData(roleOptions(id));
}

export function preloadRoles(queryClient: QueryClient) {
  return queryClient.ensureQueryData(allRoleOptions);
}
