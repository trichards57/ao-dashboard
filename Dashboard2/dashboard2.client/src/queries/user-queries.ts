import {
  useMutation,
  useQueryClient,
  UseSuspenseQueryOptions,
} from "@tanstack/react-query";
import { notFound } from "@tanstack/react-router";

export interface UserInfo {
  realName: string;
  userId: string;
  email: string;
  role: string;
  amrUsed: string;
  lastAuthenticated: string;
  otherClaims: Record<string, string>;
}

export interface UserWithRole {
  id: string;
  name: string;
  roleId: string;
  role: string;
}

export interface UserRoleUpdate {
  roleId: string;
}

export const useUpdateUser = (id: string) => {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: async (user: UserRoleUpdate) => {
      const response = await fetch(`/api/users/${id}`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(user),
      });

      if (!response.ok) {
        if (response.status === 404) {
          throw notFound();
        }
        throw new Error("Failed to update user.");
      }
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ["user"] });
    },
  });
};

export const userSettings: (
  id: string,
) => UseSuspenseQueryOptions<UserWithRole> = (id: string) => ({
  queryKey: ["user", id],
  queryFn: async () => {
    const response = await fetch(`/api/users/${id}`, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    });

    if (!response.ok) {
      if (response.status === 404) {
        throw notFound();
      }
      throw new Error("Failed to fetch user settings.");
    }

    return response.json() as Promise<UserWithRole>;
  },
  staleTime: 10 * 60 * 1000
});

export const allUserOptions: UseSuspenseQueryOptions<UserWithRole[]> = {
  queryKey: ["user"],
  queryFn: async () => {
    const response = await fetch("/api/users", {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    });

    if (!response.ok) {
      throw new Error("Failed to fetch users.");
    }

    return response.json() as Promise<UserWithRole[]>;
  },
  staleTime: 10 * 60 * 1000
};

export const userMeOptions: UseSuspenseQueryOptions<UserInfo | null> = {
  queryKey: ["me"],
  queryFn: async () => {
    const response = await fetch("/api/users/me", {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    });

    if (!response.ok) {
      return Promise.resolve(null);
    }

    return response.json() as Promise<UserInfo>;
  },
  staleTime: 1000 * 60 * 5,
};
