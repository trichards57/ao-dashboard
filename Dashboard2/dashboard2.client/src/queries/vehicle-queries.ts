import { useMutation, useQueryClient } from "@tanstack/react-query";

export interface VehicleSettings {
  id: string;
  registration: string;
  callSign: string;
  hub: string;
  district: string;
  region: string;
  vehicleType: string;
  forDisposal: boolean;
}

export interface UpdateVehicleSettings {
  registration: string;
  hub: string;
  callSign: string;
  district: string;
  region: string;
  vehicleType: string;
  forDisposal: boolean;
}

export const useUpdateVehicle = () => {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: async (vehicle: UpdateVehicleSettings) => {
      const response = await fetch("/api/vehicles", {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(vehicle),
      });

      if (!response.ok) {
        throw new Error("Failed to update vehicle.");
      }
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ["vehicle"] });
    },
  });
};

export const vehicleSettings = (id: string) => ({
  queryKey: ["vehicle", id],
  queryFn: async () => {
    const response = await fetch(`/api/vehicles/${id}`, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    });

    if (!response.ok) {
      throw new Error("Failed to fetch vehicle settings.");
    }

    return response.json() as Promise<VehicleSettings>;
  },
});

export const settingsOptions = (
  region: string | undefined,
  district: string | undefined,
  hub: string | undefined,
) => ({
  queryKey: ["settings", region ?? "All", district ?? "All", hub ?? "All"],
  queryFn: async () => {
    const response = await fetch(
      `/api/vehicles?region=${region ?? "All"}&district=${district ?? "All"}&hub=${hub ?? "All"}`,
      {
        method: "GET",
        headers: {
          "Content-Type": "application/json",
        },
      },
    );

    if (!response.ok) {
      throw new Error("Failed to fetch vehicle settings.");
    }

    return response.json() as Promise<VehicleSettings[]>;
  },
});
