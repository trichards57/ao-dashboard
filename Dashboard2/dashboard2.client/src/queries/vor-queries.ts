export interface VorStatistics {
  totalVehicles: number;
  availableVehicles: number;
  vorVehicles: number;
  pastAvailability: Record<string, number>;
}

export interface VorStatus {
  id: string;
  region: string;
  hub: string;
  district: string;
  registration: string;
  callSign: string;
  summary: string;
  isVor: boolean;
  dueBack: string;
}

export const statisticsOptions = (
  region: string,
  district: string,
  hub: string,
) => ({
  queryKey: ["statistics", region ?? "All", district ?? "All", hub ?? "All"],
  queryFn: async () => {
    const response = await fetch(
      `/api/vor/statistics?region=${region ?? "All"}&district=${district ?? "All"}&hub=${hub ?? "All"}`,
      {
        method: "GET",
        headers: {
          "Content-Type": "application/json",
        },
      },
    );

    if (!response.ok) {
      throw new Error("Failed to fetch hubs.");
    }

    return response.json() as Promise<VorStatistics>;
  },
});

export const statusOptions = (
  region: string,
  district: string,
  hub: string,
) => ({
  queryKey: ["status", region ?? "All", district ?? "All", hub ?? "All"],
  queryFn: async () => {
    const response = await fetch(
      `/api/vor?region=${region ?? "All"}&district=${district ?? "All"}&hub=${hub ?? "All"}`,
      {
        method: "GET",
        headers: {
          "Content-Type": "application/json",
        },
      },
    );

    if (!response.ok) {
      throw new Error("Failed to fetch hubs.");
    }

    return response.json() as Promise<VorStatus[]>;
  },
});
