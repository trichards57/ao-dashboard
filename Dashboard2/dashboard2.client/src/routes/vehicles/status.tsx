import { createFileRoute, redirect } from "@tanstack/react-router";
import PlacePicker from "../../components/place-picker";
import { preloadStatus, useStatus } from "../../queries/vor-queries";
import { useState } from "react";
import PagePicker from "../../components/page-picker";
import validatePlace from "../../support/validate-place";
import { useTitle } from "../../components/useTitle";
import { preloadDistricts, preloadHubs } from "../../queries/place-queries";

const PageSize = 10;

function VehicleStatus({
  region,
  district,
  hub,
}: {
  region: string | undefined;
  district: string | undefined;
  hub: string | undefined;
}) {
  const { data } = useStatus(region, district, hub);
  const [page, setPage] = useState(0);

  useTitle("Vehicle Status");

  const itemsToDisplay = (data ?? [])
    .toSorted((a, b) => a.callSign.localeCompare(b.callSign))
    .slice(page * PageSize, (page + 1) * PageSize);
  const showPagination = data.length > PageSize;

  return (
    <>
      <h1 className="title">Vehicle Status</h1>
      <PlacePicker />
      <div className="table-container">
        <table className="table is-striped is-fullwidth">
          <thead>
            <tr>
              <th className="call-sign-col">Call Sign</th>
              <th className="reg-col">Registration</th>
              <th className="vor-col">Is VOR?</th>
              <th className="due-back-col">Expected Back</th>
              <th>Summary</th>
            </tr>
          </thead>
          <tbody>
            {itemsToDisplay.length > 0 &&
              itemsToDisplay.map((i) => (
                <tr key={i.id}>
                  <td className="call-sign-col" title={i.callSign}>
                    {i.callSign}
                  </td>
                  <td className="reg-col">{i.registration}</td>
                  <td className="vor-col">
                    <span
                      className={i.isVor ? "tag is-danger" : "tag is-success"}
                    >
                      {i.isVor ? "Yes" : "No"}
                    </span>
                  </td>
                  <td className="due-back-col">{i.dueBack}</td>
                  <td className="summary" title={i.summary}>
                    {i.summary}
                  </td>
                </tr>
              ))}

            {itemsToDisplay.length === 0 && (
              <tr>
                <td colSpan={5} className="no-vehicles subtitle is-6">
                  No Vehicles
                </td>
              </tr>
            )}
          </tbody>
        </table>
      </div>
      {showPagination && (
        <PagePicker
          page={page}
          pages={Math.ceil(data.length / PageSize)}
          setPage={setPage}
        />
      )}
    </>
  );
}

export const Route = createFileRoute("/vehicles/status")({
  beforeLoad: ({ context, search }) => {
    if (!context.loggedIn) {
      throw redirect({
        to: "/",
      });
    }
    if (!context.canViewVor) {
      throw redirect({
        to: "/home",
        search: search,
      });
    }
  },
  validateSearch: validatePlace,
  loaderDeps: ({ search: { region, district, hub } }) => ({
    region,
    district,
    hub,
  }),
  loader: ({ deps, context }) =>
    Promise.all([
      preloadStatus(context.queryClient, deps.region, deps.district, deps.hub),
      preloadDistricts(context.queryClient, deps.region ?? "All"),
      preloadHubs(
        context.queryClient,
        deps.region ?? "All",
        deps.district ?? "All",
      ),
    ]),
  component: function Component() {
    return (
      <VehicleStatus
        region={Route.useSearch().region}
        district={Route.useSearch().district}
        hub={Route.useSearch().hub}
      />
    );
  },
});
