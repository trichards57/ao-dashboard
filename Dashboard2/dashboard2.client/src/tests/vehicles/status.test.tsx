import { render } from "@testing-library/react";
import { act } from "react";
import ReactDOM from "react-dom/client";
import { Mock, beforeEach, describe, expect, it, vi } from "vitest";

import { useDistricts, useHubs } from "../../queries/place-queries";
import { VorStatus, useStatus } from "../../queries/vor-queries";
import { VehicleStatus } from "../../routes/vehicles/status";

vi.mock("../../queries/vor-queries");
vi.mock("../../queries/place-queries");

const testItem: VorStatus[] = [
  {
    callSign: "WR122",
    district: "Avon",
    hub: "Bristol",
    id: "1",
    registration: "AB12 CDE",
    region: "SouthWest",
    dueBack: "2022-12-10",
    isVor: true,
    summary: "Vehicle failed",
  },
  {
    callSign: "WR123",
    district: "Avon",
    hub: "Bristol",
    id: "2",
    registration: "AB12 HTE",
    region: "SouthWest",
    dueBack: "",
    isVor: false,
    summary: "",
  },
  {
    callSign: "WR124",
    district: "Avon",
    hub: "Bristol",
    id: "3",
    registration: "AB12 CFE",
    region: "SouthWest",
    dueBack: "",
    isVor: true,
    summary: "Vehicle also failed",
  },
];

const testDistricts = ["Avon", "Bedfordshire", "Berkshire", "Buckinghamshire"];
const testHubs = ["Bristol", "Bath", "Weston-super-Mare", "Yeovil"];

describe("Vehicle Status Component", () => {
  beforeEach(() => {
    (useStatus as Mock).mockReturnValue({
      data: testItem,
    });
    (useDistricts as Mock).mockReturnValue({
      data: testDistricts,
    });
    (useHubs as Mock).mockReturnValue({
      data: testHubs,
    });
  });

  it("renders without error", () => {
    const root = ReactDOM.createRoot(document.createElement("div"));
    expect(() => {
      act(() => {
        root.render(<VehicleStatus region="All" district="All" hub="All" />);
        root.unmount();
      });
    }).not.toThrow();
  });

  it("renders correctly", () => {
    const { asFragment } = render(
      <VehicleStatus region="SouthWest" district="Avon" hub="Bristol" />,
    );
    expect(asFragment()).toMatchSnapshot();
    expect(useStatus).toHaveBeenCalledWith("SouthWest", "Avon", "Bristol");
    expect(useDistricts).toHaveBeenCalledWith("SouthWest");
    expect(useHubs).toHaveBeenCalledWith("SouthWest", "Avon");
  });

  it("renders nothing correctly", () => {
    (useStatus as Mock).mockReturnValue({
      data: [],
    });

    const { asFragment } = render(
      <VehicleStatus region="SouthWest" district="Avon" hub="Bristol" />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
