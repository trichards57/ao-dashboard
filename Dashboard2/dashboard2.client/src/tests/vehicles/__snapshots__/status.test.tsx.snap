// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Vehicle Status Component > renders correctly 1`] = `
<DocumentFragment>
  <h1
    class="title"
  >
    Vehicle Status
  </h1>
  <div
    class="columns is-desktop"
  >
    <div
      class="column"
    >
      <div
        class="field"
      >
        <label
          class="label"
          for="region"
        >
          Region
        </label>
        <div
          class="select is-fullwidth"
        >
          <select
            id="region"
          >
            <option
              value="All"
            >
              All
            </option>
            <option
              value="EastOfEngland"
            >
              East of England
            </option>
            <option
              value="EastMidlands"
            >
              East Midlands
            </option>
            <option
              value="London"
            >
              London
            </option>
            <option
              value="NorthEast"
            >
              North East
            </option>
            <option
              value="NorthWest"
            >
              North West
            </option>
            <option
              value="SouthEast"
            >
              South East
            </option>
            <option
              value="SouthWest"
            >
              South West
            </option>
            <option
              value="WestMidlands"
            >
              West Midlands
            </option>
          </select>
        </div>
      </div>
    </div>
    <div
      class="column"
    >
      <div
        class="field"
      >
        <label
          class="label"
          for="district"
        >
          District
        </label>
        <div
          class="select is-fullwidth"
        >
          <select
            id="region"
          >
            <option
              value="All"
            >
              All
            </option>
            <option
              value="Avon"
            >
              Avon
            </option>
            <option
              value="Bedfordshire"
            >
              Bedfordshire
            </option>
            <option
              value="Berkshire"
            >
              Berkshire
            </option>
            <option
              value="Buckinghamshire"
            >
              Buckinghamshire
            </option>
          </select>
        </div>
      </div>
    </div>
    <div
      class="column"
    >
      <div
        class="field"
      >
        <label
          class="label"
          for="hub"
        >
          Hub
        </label>
        <div
          class="select is-fullwidth"
        >
          <select
            id="hub"
          >
            <option
              value="all"
            >
              All
            </option>
            <option
              value="Bristol"
            >
              Bristol
            </option>
            <option
              value="Bath"
            >
              Bath
            </option>
            <option
              value="Weston-super-Mare"
            >
              Weston-super-Mare
            </option>
            <option
              value="Yeovil"
            >
              Yeovil
            </option>
          </select>
        </div>
      </div>
    </div>
  </div>
  <div
    class="table-container"
  >
    <table
      class="table is-striped is-fullwidth"
    >
      <thead>
        <tr>
          <th
            class="call-sign-col"
          >
            Call Sign
          </th>
          <th
            class="reg-col"
          >
            Registration
          </th>
          <th
            class="vor-col"
          >
            Is VOR?
          </th>
          <th
            class="due-back-col"
          >
            Expected Back
          </th>
          <th>
            Summary
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="call-sign-col"
            title="WR122"
          >
            WR122
          </td>
          <td
            class="reg-col"
          >
            AB12 CDE
          </td>
          <td
            class="vor-col"
          >
            <span
              class="tag is-danger"
            >
              Yes
            </span>
          </td>
          <td
            class="due-back-col"
          >
            2022-12-10
          </td>
          <td
            class="summary"
            title="Vehicle failed"
          >
            Vehicle failed
          </td>
        </tr>
        <tr>
          <td
            class="call-sign-col"
            title="WR123"
          >
            WR123
          </td>
          <td
            class="reg-col"
          >
            AB12 HTE
          </td>
          <td
            class="vor-col"
          >
            <span
              class="tag is-success"
            >
              No
            </span>
          </td>
          <td
            class="due-back-col"
          />
          <td
            class="summary"
            title=""
          />
        </tr>
        <tr>
          <td
            class="call-sign-col"
            title="WR124"
          >
            WR124
          </td>
          <td
            class="reg-col"
          >
            AB12 CFE
          </td>
          <td
            class="vor-col"
          >
            <span
              class="tag is-danger"
            >
              Yes
            </span>
          </td>
          <td
            class="due-back-col"
          />
          <td
            class="summary"
            title="Vehicle also failed"
          >
            Vehicle also failed
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;

exports[`Vehicle Status Component > renders nothing correctly 1`] = `
<DocumentFragment>
  <h1
    class="title"
  >
    Vehicle Status
  </h1>
  <div
    class="columns is-desktop"
  >
    <div
      class="column"
    >
      <div
        class="field"
      >
        <label
          class="label"
          for="region"
        >
          Region
        </label>
        <div
          class="select is-fullwidth"
        >
          <select
            id="region"
          >
            <option
              value="All"
            >
              All
            </option>
            <option
              value="EastOfEngland"
            >
              East of England
            </option>
            <option
              value="EastMidlands"
            >
              East Midlands
            </option>
            <option
              value="London"
            >
              London
            </option>
            <option
              value="NorthEast"
            >
              North East
            </option>
            <option
              value="NorthWest"
            >
              North West
            </option>
            <option
              value="SouthEast"
            >
              South East
            </option>
            <option
              value="SouthWest"
            >
              South West
            </option>
            <option
              value="WestMidlands"
            >
              West Midlands
            </option>
          </select>
        </div>
      </div>
    </div>
    <div
      class="column"
    >
      <div
        class="field"
      >
        <label
          class="label"
          for="district"
        >
          District
        </label>
        <div
          class="select is-fullwidth"
        >
          <select
            id="region"
          >
            <option
              value="All"
            >
              All
            </option>
            <option
              value="Avon"
            >
              Avon
            </option>
            <option
              value="Bedfordshire"
            >
              Bedfordshire
            </option>
            <option
              value="Berkshire"
            >
              Berkshire
            </option>
            <option
              value="Buckinghamshire"
            >
              Buckinghamshire
            </option>
          </select>
        </div>
      </div>
    </div>
    <div
      class="column"
    >
      <div
        class="field"
      >
        <label
          class="label"
          for="hub"
        >
          Hub
        </label>
        <div
          class="select is-fullwidth"
        >
          <select
            id="hub"
          >
            <option
              value="all"
            >
              All
            </option>
            <option
              value="Bristol"
            >
              Bristol
            </option>
            <option
              value="Bath"
            >
              Bath
            </option>
            <option
              value="Weston-super-Mare"
            >
              Weston-super-Mare
            </option>
            <option
              value="Yeovil"
            >
              Yeovil
            </option>
          </select>
        </div>
      </div>
    </div>
  </div>
  <div
    class="table-container"
  >
    <table
      class="table is-striped is-fullwidth"
    >
      <thead>
        <tr>
          <th
            class="call-sign-col"
          >
            Call Sign
          </th>
          <th
            class="reg-col"
          >
            Registration
          </th>
          <th
            class="vor-col"
          >
            Is VOR?
          </th>
          <th
            class="due-back-col"
          >
            Expected Back
          </th>
          <th>
            Summary
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="no-vehicles subtitle is-6"
            colspan="5"
          >
            No Vehicles
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
